src/pymoca/tree.py

```diff
@@ -259,6 +273,694 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+    assert isinstance(name, (str, ast.ComponentRef))
+    if isinstance(name, str):
+        left_name, _, rest_of_name = name.partition(".")
+    else:
+        name_parts = name.to_tuple()
+        left_name = name_parts[0]
+        rest_of_name = ".".join(name_parts[1:])
+    return left_name, rest_of_name
+
+
+def _find_simple_name(
+    name: str,
+    scope: ast.Class,
+    search_imports: bool = True,
+    search_parent: bool = True,
+    check_builtin_classes: bool = False,
+    current_extends: Optional[Set[ast.ExtendsClause]] = None,
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Lookup name per Modelica spec 3.5 section 5.3.1 Simple Name Lookup
+
+    0. Predefined types (`Real`, `Integer`, `Boolean`, `String`) (spec 4.8)
+    1. Iteration variables
+    2. Classes
+    3. Components (Symbols in Pymoca)
+    4. Classes and Components from Extends Clauses
+    5. Qualified Import names, see 4 (but not from Extends Clauses) (spec 13.2.1)
+    6. Public Unqualified Imports (error if multiple are found) (spec 13.2.1)
+    7. Repeat 1-6 for each lexically enclosing instance scope, stopping at `encapsulated`
+    unless predefined type, function, operator. If name matches a variable (a.k.a. component,
+    symbol) in an enclosing class, it must be a `constant`.
+    """
+
+    current_scope = scope
+    while True:
+        if (
+            (
+                found := _find_local(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                )
+            )
+            or (
+                found := _find_inherited(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                    current_extends=current_extends,
+                )
+            )
+            or search_imports
+            and (
+                found := _find_imported(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                    current_extends=current_extends,
+                )
+            )
+            or not search_parent
+            or not current_scope.parent
+            or current_scope.encapsulated
+        ):
+            break
+        current_scope = current_scope.parent
+    # If name matches a variable (a.k.a. component a.k.a. symbol) in an enclosing class,
+    # it must be a `constant`.
+    if (
+        isinstance(found, ast.Symbol)
+        and current_scope != scope
+        and "constant" not in found.prefixes
+    ):
+        raise NameLookupError("Non-constant Symbol found in enclosing class")
+
+    # If not found and we stopped at an encapsulated class,
+    # then search predefined functions and operators in global scope
+    # TODO: Add predefined functions and operators to global scope before this
```

Comment:
I have not checked, but I believe, per the grammar, if a user tries to define something at the top level it will be a syntax error. The parser should enforce this error.